lib/auth/predicate/partial/solver.go

```diff
@@ -0,0 +1,323 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package predicate
+
+import (
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/aclements/go-z3/z3"
+	"github.com/gravitational/trace"
+)
+
+type Resolver func([]string) any
+type Type int
+
+const (
+	TypeInt Type = iota
+	TypeBool
+	TypeString
+)
+
+type CachedSolver struct {
+	def    *z3.Context
+	solver *z3.Solver
+}
+
+func NewCachedSolver() *CachedSolver {
+	config := z3.NewContextConfig()
+	def := z3.NewContext(config)
+	solver := z3.NewSolver(def)
+	return &CachedSolver{def, solver}
+}
+
+func (s *CachedSolver) PartialSolveForAll(predicate string, resolveIdentifier Resolver, querying string, to Type, timeout time.Duration) ([]z3.Value, error) {
+	outCh := make(chan []z3.Value)
+	errCh := make(chan error)
+
+	go func() {
+		out, err := s.partialSolveForAllImpl(predicate, resolveIdentifier, querying, to)
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		outCh <- out
```

Comment:
since these channels are unbuffered i believe this goroutine will hang/leak if you exit due to timeout

Reply:
fixed by buffering once